ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionSimulator.java

```diff
@@ -0,0 +1,203 @@
+              0L,
+              ValidationResult.invalid(
+                  TransactionValidator.TransactionInvalidReason.PRIVACY_GROUP_DOES_NOT_EXIST),
+              Optional.empty()));
+    }
+
+    final MutableWorldState publicWorldState =
+        worldStateArchive.getMutable(header.getStateRoot()).orElse(null);
+    if (publicWorldState == null) {
+      return Optional.empty();
+    }
+
+    // get the last world state root hash or create a new one
+    final Bytes privacyGroupId = Bytes.fromBase64String(privacyGroupIdString);
+    final Hash lastRootHash =
+        privacyParameters
+            .getPrivateStateStorage()
+            .getLatestStateRoot(privacyGroupId)
+            .orElse(EMPTY_ROOT_HASH);
+
+    final MutableWorldState disposablePrivateState =
+        privacyParameters.getPrivateWorldStateArchive().getMutable(lastRootHash).get();
+
+    final Address senderAddress =
+        callParams.getFrom() != null ? callParams.getFrom() : DEFAULT_FROM;
+    final Account sender = publicWorldState.get(senderAddress);
+    final long nonce = sender != null ? sender.getNonce() : 0L;
+    final long gasLimit =
+        callParams.getGasLimit() >= 0 ? callParams.getGasLimit() : header.getGasLimit();
+    final Wei gasPrice = callParams.getGasPrice() != null ? callParams.getGasPrice() : Wei.ZERO;
+    final Wei value = callParams.getValue() != null ? callParams.getValue() : Wei.ZERO;
+    final Bytes payload = callParams.getPayload() != null ? callParams.getPayload() : Bytes.EMPTY;
+
+    final PrivateTransaction transaction =
+        PrivateTransaction.builder()
+            .privateFrom(Bytes.EMPTY)
+            .privacyGroupId(privacyGroupId)
+            .restriction(Restriction.RESTRICTED)
+            .nonce(nonce)
+            .gasPrice(gasPrice)
+            .gasLimit(gasLimit)
+            .to(callParams.getTo())
+            .sender(senderAddress)
+            .value(value)
+            .payload(payload)
+            .signature(FAKE_SIGNATURE)
+            .build();
+
+    final ProtocolSpec<?> protocolSpec = protocolSchedule.getByBlockNumber(header.getNumber());
+
+    final PrivateTransactionProcessor privateTransactionProcessor =
+        protocolSpec.getPrivateTransactionProcessor();
+
+    final PrivateTransactionProcessor.Result result =
+        privateTransactionProcessor.processTransaction(
+            blockchain,
+            publicWorldState.updater(),
+            disposablePrivateState.updater(),
+            header,
+            transaction,
+            protocolSpec.getMiningBeneficiaryCalculator().calculateBeneficiary(header),
+            new DebugOperationTracer(TraceOptions.DEFAULT),
+            new BlockHashLookup(header, blockchain),
+            privacyGroupId);
+
+    return Optional.of(result);
+  }
+
+  public Optional<Boolean> doesAddressExistAtHead(final Address address) {
+    return doesAddressExist(address, blockchain.getChainHeadHeader());
+  }
```

Comment:
Looks like this method isn't being used.

Reply:
removed